File: ts/rc/lsp.ts
import * as lsp from '../lsp/index';
import { serverCfgs } from './lspconfigdata';
import { map } from '../utils';
import { autocmd } from '../utils/autocmd';

autocmd({
  id: 'StartServer',
  events: 'BufRead',
  filetypes: '*',
  command: ':LspStart'
});

autocmd({
  id: 'LspFormat',
  events: 'BufWritePre',
  filetypes: '*',
  command: () => {
    vim.lsp.buf.formatting_sync();
  }
});
map.nmap('<leader>f', ':lua vim.lsp.buf.formatting()<CR>', map.ns);

lsp.setup();

lsp.ensureInstallServers([
  'cssls',
  'html',
  'jsonls',
  'tsserver',
  'diagnosticls',
  'svelte',
  'vimls',
  'vuels',
  'yamlls',
  'bashls',
  'tailwindcss',
  'rust_analyzer'
]);

const manualServers = ['denols'];

const manualServersSetup = () => {
  const servers = manualServers;
  servers.forEach((serverName) => {
    const cfg: NoColonType = (serverCfgs as any)[serverName];
    (require('lspconfig')[serverName] as any as NoColonType).setup({
      ...cfg,
      on_attach: () => {
        require('lsp_signature').on_attach();
      }
    });
  });
};
const setup = () => {
  const servers = lsp.getInstalledServers();
  if (servers.includes('diagnosticls')) {
    servers.push(servers.splice(servers.indexOf('diagnosticls'), 1)[0]);
  }
  servers.forEach((serverName) => {
    const cfg: NoColonType = (serverCfgs as any)[serverName];
    lsp.setupServer(serverName, {
      ...cfg,
      on_attach: (client: any) => {
        require('lsp_signature').on_attach();
        cfg.on_attach ? cfg.on_attach(client) : '';
      }
    });
  });
};

manualServersSetup();
setup();

File: lua/rc/lsp.lua
local autocmd = require 'utils.autocmd'
local map = require 'utils.map'

local inspect = require 'lib.inspect'

local lspConfigs = require 'rc.lspconfigs'

local function setup_servers()
  require('lspinstall').setup()
  local servers = require('lspinstall').installed_servers()
  for _, server in pairs(servers) do
    local config = lspConfigs[server]
    if config == nil then
      config = {}
    else
      -- print(inspect(config))
    end
    require('lspconfig')[server].setup(config)
  end
end

setup_servers()

-- Automatically reload after `:LspInstall <server>` so we don't have to restart neovim
require('lspinstall').post_install_hook = function()
  setup_servers() -- reload installed servers
  vim.cmd 'bufdo e'
end

local nullls = require 'null-ls'
nullls.config {
  sources = {
    nullls.builtins.formatting.stylua,
    nullls.builtins.formatting.eslint_d,
  },
}
require('lspconfig')['null-ls'].setup {}

autocmd {
  id = 'LspFormat',
  events = 'BufWritePre',
  filetypes = '*',
  command = function()
    vim.lsp.buf.formatting_sync()
  end,
}
map.nmap('<leader>f', ':lua vim.lsp.buf.formatting()<CR>', map.ns)
